Jv investigate getting processindicatorid without db lookup

JoukoVirtanen opened the pull request:
## Description

Instead of looking up processindicatorid in the process_indicator table using the process information in the plop object, determine the processindicatorid using the process information in the plop object using the same function that is used to determine the id for the process indicators. This is both more efficient and means that we get process information for listening endpoints when the listening endpoint arrives before the process indicator, without any complex mechanism for handling that case.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Ran e2e tests 200 times
- [x] Load testing
- [x] Performance testing

## Testing Performed

To test this properly the changes here were merged with other fixes. A PR for testing purposes was created here https://github.com/stackrox/stackrox/pull/5393

Performance tests are here https://docs.google.com/document/d/1Xv8eKZWiXLURGn9LVrLdo9o4-WJR49A0f7thInlBAFY/edit#heading=h.aptf438z8hjb

Load tests are here https://docs.google.com/document/d/1THsA2PohUYO_SYTT4uL2VcLmBKvVikW35x70nK9WP0E/edit#heading=h.wmfol0u90wc9

JoukoVirtanen commented:
/test ocp-4-10-qa-e2e-tests

dashrews78 commented on pkg/postgres/schema/process_indicators.go:
```diff
@@ -58,13 +58,13 @@ const (
 type ProcessIndicators struct {
```
Just to clarify something to make sure we do not need a migration.  If memory serves these updates were NOT part of 3.74.  Is that correct?  If that is the case we do not need a migration to remove those indexes.

JoukoVirtanen replied:
Nothing related to processes listening on ports went into central for 3.74, except the feature flag. These updates were not a part of 3.74.

erthalion reviewed (commented):
Few comments above.

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -116,10 +120,10 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
 			}
 
-			plopObjects = addNewPLOP(plopObjects, indicatorID, processInfo, val)
+			plopObjects = addNewPLOP(plopObjects, indicatorID, nil, val)
```
I brought in your commit from the other branch to fix this. There were a couple of minor mistakes in it that I fixed. Fix is confirmed by the unit tests.

JoukoVirtanen commented:
/test ocp-4-8-qa-e2e-tests
/test ocp-4-10-qa-e2e-tests

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -63,40 +85,44 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
-			processInfo = val.Process
-		}
-
+		indicatorID := getIndicatorIDForPlop(val)
```
Do you need to handle the case where this call returns an empty string?  Or does processing work as desired in that case?

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -130,24 +156,21 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
-			processInfo = val.Process
-		}
-
+		indicatorID := getIndicatorIDForPlop(val)
```
same comment above as this returning empty and do we need to handle that case or not.

erthalion commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -63,40 +85,44 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
+	allPLOPs := append(normalizedPLOPs, completedInBatch...)
+	indicatorIds := getIndicatorIdsForPlops(allPLOPs)
+
+	indicators, nonempty, err := ds.indicatorDataStore.GetProcessIndicators(ctx, indicatorIds)
```
`err == nil` is used a few lines below to actually prepare `indicatorsMap`. Is that what you asking about?

dashrews78 replied:
yeah but we aren't doing anything if that call returns an `err != nil`.  Maybe we want to treat the case of `ds.indicatorDataStore.GetProcessIndicators` return an error the same as if it returns nothing or maybe we want to do something different when `err != nil`.  I was asking to make sure not handling `err != nil` was intentional.

erthalion replied:
Right, sorry for not communicating it more clearly. The idea was to treat it in the same way as empty response, not preparing `indicatorsMap` in such case. As this logic is purely for reducing amount of data we duplicate, every error scenario could be considered as not fatal, switch us to less performant path.

dashrews78 replied:
got it.  That clarification makes sense to me.

JoukoVirtanen replied:
So no changes are needed here?

dashrews78 replied:
@JoukoVirtanen correct.  fine as is.

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -40,6 +40,28 @@ func newDatastoreImpl(
 }
 
+func getIndicatorIDForPlop(plop *storage.ProcessListeningOnPortFromSensor) string {
+	if plop == nil {
```
Now it is there.

JoukoVirtanen replied:
Improve warning to say what the consequence of not being able to set id is.

JoukoVirtanen commented:
/test ocp-4-10-qa-e2e-tests

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -7,14 +7,14 @@ import (
 	"time"
 
 	"github.com/stackrox/rox/central/metrics"
-	countMetrics "github.com/stackrox/rox/central/metrics"
```
Add metric back in

JoukoVirtanen replied:
Done

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -486,7 +517,7 @@ func addNewPLOP(plopObjects []*storage.ProcessListeningOnPortStorage,
 	value *storage.ProcessListeningOnPortFromSensor) []*storage.ProcessListeningOnPortStorage {
 
-	if value == nil {
+	if value == nil || indicatorID == "" {
```
log.Warn here

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -63,40 +87,47 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
+		}
 	}
 
+	// TODO ROX-14376: This ticket might need to be changed, but is not completely obsolete
```
Check ticket and see if it is still needed.

JoukoVirtanen replied:
I don't think it is needed. I deleted the comment and closed the ticket.

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl_test.go:
```diff
@@ -134,6 +133,20 @@ func (suite *PLOPDataStoreTestSuite) TestPLOPAdd() {
 		},
 	}
 
+	for _, indicator := range indicators {
```
Multiple events with no matching process indicators. Two consecutive open and two consecutive close. Enable unit tests in CI.

JoukoVirtanen replied:
Done

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -63,40 +87,47 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
+		// need to store the process information. Otherwise processInfo is nil
+		// and will not be stored.
+		if _, indicatorExists := indicatorsMap[indicatorID]; !indicatorExists {
+			processInfo = val.GetProcess()
```
Add metric here.

JoukoVirtanen replied:
Done

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -130,24 +161,24 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
+		// need to store the process information. Otherwise processInfo is nil
+		// and will not be stored.
+		if _, indicatorExists := indicatorsMap[indicatorID]; !indicatorExists {
+			processInfo = val.GetProcess()
```
Done

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -63,40 +87,47 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
-			processInfo = val.Process
+		indicatorID := getIndicatorIDForPlop(val)
+		if indicatorID == "" {
+			continue
```
Done

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -130,24 +161,24 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
-			processInfo = val.Process
+		indicatorID := getIndicatorIDForPlop(val)
+		if indicatorID == "" {
+			continue
```
Done

JoukoVirtanen committed:
X-Smart-Squash: Squashed 3 commits:

f6cf0c3aa0 Generating processindicatorid from plop info instead of doing join
095d35bed4 Cleaned up
af2f486da5 Fixed style errors. Reenabled e2e tests

JoukoVirtanen committed:
X-Smart-Branch-Parent: master

JoukoVirtanen committed:
X-Smart-Squash: Squashed 7 commits:

ec116e1311 Check scale performance with indicator index off
146a1cce12 Add best efforts ProcessIndicator matching
2cb006fb12 Setting process correctly
8e39321b58 Check if indicatorID is empty in addNewPLOP
75865b0636 Addressed PR review comments. Changed the type of the value of indicatorsMap to bool since it is not being used
4e25c7d19b Added back metric for orphaned plop
abaebbdb9a Added tests. Added logging. Updated comments

JoukoVirtanen committed:
X-Smart-Branch-Parent: jv-investigate-getting-processindicatorid-without-join

JoukoVirtanen committed:
README.md

JoukoVirtanen committed:
Minor change to comment

JoukoVirtanen committed:
Fixed style error

JoukoVirtanen committed:
X-Smart-Branch-Parent: jv-rox-15995-plop-service-should-report-endpoint-when-process-indicator-is-not-found

JoukoVirtanen committed:
Empty commit to trigger ci

erthalion reviewed (approved):
Please update the PR description and the squashed commit message to describe the rationale behind this change.

JoukoVirtanen committed:
Jv rox 15995 plop service should report endpoint when process indicator is not found (#5265)

JoukoVirtanen committed:
Fixed style

JoukoVirtanen committed:
Fixed style

JoukoVirtanen committed:
Undid unneeded change to updating deploymentid for plop

JoukoVirtanen committed:
Disabled e2e tests

JoukoVirtanen commented:
/test ocp-4-12-qa-e2e-tests

JoukoVirtanen commented:
/test gke-nongroovy-e2e-tests
/test gke-postgres-nongroovy-e2e-tests

JoukoVirtanen commented:
/test gke-postgres-qa-e2e-tests
/test ocp-4-8-qa-e2e-tests
/test gke-qa-e2e-tests
